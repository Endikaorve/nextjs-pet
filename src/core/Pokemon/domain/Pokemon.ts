export interface Pokemon {
  id: string;
  name: string;
  height: number;
  weight: number;
  types: string[];
  img: string;
  alt_img: string;
  description: string;
}
