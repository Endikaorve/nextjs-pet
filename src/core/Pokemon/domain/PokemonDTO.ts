import { ColorType } from "styles/theme/Colors";

export interface PokemonDTO {
  id: string;
  name: string;
  height: number;
  weight: number;
  types: {
    slot: number;
    type: {
      name: ColorType;
      url: string;
    };
  }[];
  sprites: {
    other: {
      "official-artwork": {
        front_default: string;
      };
      dream_world: {
        front_default: string;
      };
    };
  };
}
