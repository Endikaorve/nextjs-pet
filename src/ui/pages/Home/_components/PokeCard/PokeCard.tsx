import Image from "next/image";
import Link from "next/link";

import styled from "styled-components";
import { Color, ColorType } from "styles/theme/Colors";

import { Pokemon } from "src/core/Pokemon/domain/Pokemon";

interface Props {
  pokemon: Pokemon;
}

const PokeCard: React.FC<Props> = ({ pokemon }) => (
  <Link href={`/details/${pokemon.name}`}>
    <a>
      <Card type={pokemon.types[0]}>
        <CardHeader>
          <CardHeaderName>{pokemon.name}</CardHeaderName>
          <CardHeaderCode>#{pokemon.id}</CardHeaderCode>
        </CardHeader>
        <CardBody>
          <CardBodyImgWrapper>
            <Image
              src={pokemon.altImg}
              alt={pokemon.id}
              layout={"fill"}
              objectFit={"contain"}
            />
          </CardBodyImgWrapper>

          <CardBodyTypes>
            {pokemon.types.map((type: any) => (
              <CardBodyTypesBadge
                key={`type_badge_${pokemon.id}_${type}`}
                type={type}
              >
                {type}
              </CardBodyTypesBadge>
            ))}
          </CardBodyTypes>
          <CardBodyAbout>About</CardBodyAbout>
          <CardBodyDetails>
            <CardBodyDetailsContent>
              <CardBodyDetailsContentData>
                {/* <CardBodyDetailsContentDataImg src={weight} alt="weight-icon" /> */}
                {pokemon.weight} kg
              </CardBodyDetailsContentData>
              <CardBodyDetailsContentTitle>Weight</CardBodyDetailsContentTitle>
            </CardBodyDetailsContent>

            <CardBodyDetailsContent>
              <CardBodyDetailsContentData>
                {/* <img
              className="poke-card-body-details-content-data-img"
              src={height}
              alt="height-icon"
            /> */}
                {pokemon.height} m
              </CardBodyDetailsContentData>
              <CardBodyDetailsContentTitle>Height</CardBodyDetailsContentTitle>
            </CardBodyDetailsContent>
          </CardBodyDetails>
          <CardBodyDescription>{pokemon.description}</CardBodyDescription>
        </CardBody>
      </Card>
    </a>
  </Link>
);

const Card = styled.div<{
  type: ColorType;
}>`
  position: relative;
  flex-shrink: 0;
  width: 300px;
  height: 400px;
  margin: 0 12px 24px 12px;
  border-radius: 12px;
  box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.25);

  background-color: ${({ type }) => Color[type]};
`;

const CardHeader = styled.div`
  display: flex;
  align-items: center;
  padding: 24px 24px 0px 24px;
  color: white;
  font-weight: 700;
`;

const CardHeaderName = styled.div`
  margin-right: auto;
  font-size: 24px;
  text-transform: capitalize;
`;

const CardHeaderCode = styled.div`
  font-size: 12px;
`;

const CardBody = styled.div`
  position: absolute;
  bottom: 5px;
  left: 5px;
  right: 5px;
  height: calc(100% - 160px);
  /*height: 240px;*/
  background-color: white;
  border-radius: 12px;
`;

const CardBodyImgWrapper = styled.div`
  position: absolute;
  bottom: calc(100% - 49px);
  left: 50%;
  transform: translate(-50%, 0);
  height: 140px;
  width: 250px;
`;

const CardBodyTypes = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 56px;
`;

const CardBodyTypesBadge = styled.span<{
  type: ColorType;
}>`
  display: flex;
  align-items: center;
  height: 20px;
  margin-right: 16px;
  padding: 0 10px;
  border-radius: 100px;
  color: white;
  font-size: 10px;
  line-height: 16px;
  font-weight: 700;
  text-transform: capitalize;

  background-color: ${({ type }) => Color[type]};

  &:last-child {
    margin-right: 0;
  }
`;

const CardBodyAbout = styled.div`
  margin-top: 16px;
  margin-bottom: 16px;
  text-align: center;
  font-size: 14px;
  line-height: 16px;
  font-weight: 700;
`;

const CardBodyDetails = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 16px;
`;

const CardBodyDetailsContent = styled.div`
  padding: 0 24px;
  border-right: 1px solid #e0e0e0;

  &:last-child {
    border-right: none;
  }
`;

const CardBodyDetailsContentData = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 10px;
  font-size: 10px;
  line-height: 16px;
`;

const CardBodyDetailsContentDataImg = styled.img`
  height: 16px;
  width: 16px;
  margin-right: 8px;
`;

const CardBodyDetailsContentTitle = styled.div`
  text-align: center;
  font-size: 8px;
  line-height: 12px;
  color: #666666;
`;

const CardBodyDescription = styled.div`
  height: 48px;
  overflow: hidden;
  padding: 0 20px;
  text-align: justify;
  font-size: 10px;
  line-height: 16px;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
`;

export default PokeCard;
