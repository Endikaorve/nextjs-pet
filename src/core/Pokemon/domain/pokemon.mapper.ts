import { PokemonDTO } from "./PokemonDTO";
import { Pokemon } from "./Pokemon";

const map = (pokemon: PokemonDTO): Pokemon => ({
  id: pokemon.id,
  name: pokemon.name,
  height: pokemon.height / 10,
  weight: pokemon.weight / 10,
  types: pokemon.types.map((typeContainer) => typeContainer.type.name),
  img: pokemon.sprites.other["official-artwork"].front_default,
  alt_img: pokemon.sprites.other.dream_world.front_default,
  description:
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla sit amet facilisis mi. Integer eget urna eu mauris scelerisque ornare. Nam at finibus purus. Nam quis mauris elementum, ullamcorper sem non, convallis mi. Sed a bibendum sapien, non posuere ligula. Morbi sit amet maximus purus. Nulla eu pulvinar ante. Nulla gravida luctus convallis. Duis suscipit vel nisi et tincidunt. Aenean velit purus, rutrum euismod justo id, porta tincidunt orci. Curabitur quis nibh elit. Etiam ornare tortor ac nibh commodo dapibus. Proin vehicula erat sit amet lacus posuere tincidunt.",
});

const PokemonMapper = {
  map,
};

export default PokemonMapper;
