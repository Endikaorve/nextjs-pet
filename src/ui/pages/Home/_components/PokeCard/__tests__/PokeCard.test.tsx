import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

import { Pokemon } from "src/core/Pokemon/domain/Pokemon";
import PokeCard from "../PokeCard";

const fakePokemon: Pokemon = {
  id: "1",
  name: "Bulbasaur",
  height: 6.9,
  weight: 0.7,
  types: ["grass", "poison"],
  img: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
  altImg:
    "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
  description: "Lorem ipsum dolor sit emet",
};

// DUDAS: Â¿1 solo test que compruebe todo o un test por cada campo a comprobar?

describe("PokeCard", () => {
  beforeEach(() => {
    render(<PokeCard pokemon={fakePokemon} />);
  });

  it("renders the complete card", () => {
    const pokemonName = screen.getByText(fakePokemon.name);
    const pokemonID = screen.getByText(`#${fakePokemon.id}`);
    const pokemonImage = screen.getByRole("img", {
      name: fakePokemon.id,
    });
    const pokemonWeight = screen.getByText(`${fakePokemon.weight} kg`);
    const pokemonHeight = screen.getByText(`${fakePokemon.height} m`);
    const pokemonDescription = screen.getByText(fakePokemon.description);

    expect(pokemonName).toBeInTheDocument();
    expect(pokemonID).toBeInTheDocument();
    expect(pokemonImage).toBeInTheDocument();
    expect(pokemonHeight).toBeInTheDocument();
    expect(pokemonWeight).toBeInTheDocument();
    expect(pokemonDescription).toBeInTheDocument();
  });

  it("renders the name", () => {
    const pokemonName = screen.getByText(fakePokemon.name);

    expect(pokemonName).toBeInTheDocument();
  });

  it("renders the ID", () => {
    const pokemonID = screen.getByText(`#${fakePokemon.id}`);

    expect(pokemonID).toBeInTheDocument();
  });

  it("renders the image", () => {
    const pokemonImage = screen.getByRole("img", {
      name: fakePokemon.id,
    });

    expect(pokemonImage).toBeInTheDocument();
  });
});

describe("PokeCard Snapshot", () => {
  it("Snapshots", () => {
    const pokeCard = renderer
      .create(<PokeCard pokemon={fakePokemon} />)
      .toJSON();
    expect(pokeCard).toMatchSnapshot();
  });
});
