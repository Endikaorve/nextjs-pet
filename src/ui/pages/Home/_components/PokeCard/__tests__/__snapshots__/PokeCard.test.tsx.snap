// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeCard Snapshot Snapshots 1`] = `
<a
  href="/details/Bulbasaur"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="sc-1ce23569-0 eyUTxi"
    type="grass"
  >
    <div
      className="sc-1ce23569-1 fQXFVc"
    >
      <div
        className="sc-1ce23569-2 fKGLHe"
      >
        Bulbasaur
      </div>
      <div
        className="sc-1ce23569-3 cpNnca"
      >
        #
        1
      </div>
    </div>
    <div
      className="sc-1ce23569-4 haWEmS"
    >
      <div
        className="sc-1ce23569-5 iAcChI"
      >
        <span
          style={
            Object {
              "background": "none",
              "border": 0,
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": "initial",
              "left": 0,
              "margin": 0,
              "opacity": 1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": "initial",
            }
          }
        >
          <img
            alt="1"
            data-nimg="fill"
            decoding="async"
            onError={[Function]}
            onLoad={[Function]}
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": "contain",
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
          <noscript>
            <img
              alt="1"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              sizes="100vw"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
              srcSet="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 640w, https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 750w, https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 828w, https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 1080w, https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 1200w, https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 1920w, https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 2048w, https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg 3840w"
              style={
                Object {
                  "border": "none",
                  "bottom": 0,
                  "boxSizing": "border-box",
                  "display": "block",
                  "height": 0,
                  "left": 0,
                  "margin": "auto",
                  "maxHeight": "100%",
                  "maxWidth": "100%",
                  "minHeight": "100%",
                  "minWidth": "100%",
                  "objectFit": "contain",
                  "objectPosition": undefined,
                  "padding": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 0,
                }
              }
            />
          </noscript>
        </span>
      </div>
      <div
        className="sc-1ce23569-6 cYKHPW"
      >
        <span
          className="sc-1ce23569-7 bJPhlh"
          type="grass"
        >
          grass
        </span>
        <span
          className="sc-1ce23569-7 kLnbQk"
          type="poison"
        >
          poison
        </span>
      </div>
      <div
        className="sc-1ce23569-8 eZSuxQ"
      >
        About
      </div>
      <div
        className="sc-1ce23569-9 cLwqfx"
      >
        <div
          className="sc-1ce23569-10 cvLBdK"
        >
          <div
            className="sc-1ce23569-11 fcMZuV"
          >
            0.7
             kg
          </div>
          <div
            className="sc-1ce23569-13 efIxYo"
          >
            Weight
          </div>
        </div>
        <div
          className="sc-1ce23569-10 cvLBdK"
        >
          <div
            className="sc-1ce23569-11 fcMZuV"
          >
            6.9
             m
          </div>
          <div
            className="sc-1ce23569-13 efIxYo"
          >
            Height
          </div>
        </div>
      </div>
      <div
        className="sc-1ce23569-14 dNFOkl"
      >
        Lorem ipsum dolor sit emet
      </div>
    </div>
  </div>
</a>
`;
